### ISSUE TITLE
Collapsing/Aliasing labels to a hostname?

### ISSUE THREAD

**Issue by polarathene** [openshift/coredns-mdns#62]
`coredns-mdns` discovers on the network `hostname.local` and this works well. Is there a way to resolve an additional labels such as `service.hostname.local` to `hostname.local`?

I have noticed that the `rewrite` plugin for CoreDNS can help with this if you're explicit, but [presently lacks the ability to work in a wildcard capacity](https://github.com/coredns/coredns/issues/2389#issuecomment-683241728). 

The intention is for a system to only expose it's hostname over mDNS which is working well, but redirect additional labels (which tend to be treated as Unicast from some clients, or always for Android apps that haven't had the developer explicitly configure mDNS support). 

These could then be used with a reverse-proxy and not require explicit publishing (which this plugin seems to require, if a system isn't discoverable via `avahi-browse -a` but is reachable via `avahi-resolve`, the plugin won't make it accessible), with mDNS publishers/servers like Avahi, creating additional FQDNs that map to the same IP is not all that well supported, separate from the related issue of typical mDNS two label (hostname.local) rule most clients use that advises services(as in additional labels, not DNS-SD services exposed over mDNS) to be specified at the same level as hostnames instead.

**Comment by cybertron** [comment]
So you want to publish an mdns record for hostname.local and have any requests for *.hostname.local also resolve to that address? I'm sure that could be done. If the requested name isn't found directly, you would strip the first element of the address and try the lookup again with that name.

I don't know whether that's something I'll have time to implement though. We solved the multiple names resolving to a single address problem by writing our own mdns publishing service that handles it fine: https://github.com/openshift/mdns-publisher

**Comment by polarathene** [comment]
I believe when I created the issue I wasn't aware that the discovery issues were due to a lack of publishing a hostname with a service that the plugin (or `avahi-browse`) would pick up on. That was still an issue with Avahi at least for publishing hostnames to the same host IP. 

Most mDNS clients will ignore sub-label queries (eg `service.hostname.local`) and forward to unicast. Which is a drawback/misunderstanding from the original intention I had to leverage mDNS host discovery for reaching services via a reverse-proxy on a local network without needing each client device configured to query DNS from a local DNS service such as CoreDNS (or maintaining hard-coded `hosts`). 

It is still somewhat beneficial in that it reduces static config of IPs for each host for the DNS service, so aliasing sub-labels that have no match to the hostname similar to wildcard CNAME record sounds good?

> So you want to publish an mdns record for hostname.local and have any requests for *.hostname.local also resolve to that address? I'm sure that could be done.

Yes, like regular DNS with CNAME wildcard if no explicit record exists.

> I don't know whether that's something I'll have time to implement though. 

That's fine, as it's not compatible with regular mDNS clients, the advantages isn't as attractive as I had hoped for when looking into mDNS.

> We solved the multiple names resolving to a single address problem by writing our own mdns publishing service that handles it fine: https://github.com/openshift/mdns-publisher

That looks like a great solution :+1: 

My scenario involved a web app with subdomains, if `service-hostname.local` would have been suitable I definitely would have looked into `mdns-publisher`.

---

I'll close this issue. It's rather niche with only slight improvement over unicast DNS for what I was wanting to do.
